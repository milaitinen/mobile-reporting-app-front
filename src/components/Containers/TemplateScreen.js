import React, { Component } from 'react';
import {
    StyleSheet,
    View,
    FlatList,
    Platform,
    ActivityIndicator,
    ScrollView
} from 'react-native';
import { ListItem } from 'react-native-elements';

import Panel from './Panel';
import { url } from './urlsetting';

class TemplateScreen extends Component {
    static displayName = 'TemplateScreen';

    constructor(props)
    {
        super(props);
        this.state = {
            isLoading: true,
            refreshing: false,
            switch: true,
        };
    }

    componentDidMount() {

        this.getLayoutsAndForms();

    }

    getLayouts() {

        fetch(url + '/layouts')
            .then((response) => response.json())
            .then((responseJson) => {
                this.setState({
                    isLoading: false,
                    refreshing: false,
                    dataLayouts: responseJson
                }, function() {
                    // In this block you can do something with new state.
                });
            })
            .catch((error) => {
                console.error(error);
            });

    }

    getForms() {

        fetch(url + '/forms')
            .then((response) => response.json())
            .then((responseJson) => {
                this.setState({
                    isLoading: false,
                    refreshing: false,
                    dataForms: responseJson
                }, function() {
                    // In this block you can do something with new state.
                });
            })
            .catch((error) => {
                console.error(error);
            });

    }

    getLayoutsAndForms() {
        return Promise.all([this.getLayouts(), this.getForms()]);
    }


    handleRefresh = () => {
        this.setState(
            {
                refreshing: true,
            },
            () => {
                this.componentDidMount();
            }
        );

    }

    onRowPress = () => {
        this.props.navigation.navigate('NewForm', { refresh: this.handleRefresh });
    }


    render() {

        if (this.state.isLoading) {
            return (
                <View style={[styles.container]}>

                    <ActivityIndicator
                        animating={this.state.animating}
                        style={[styles.activityIndicator, { height: 80 }]}
                        size='large'
                    />

                </View>
            );
        }

        return (
            <View style={{ flex: 1 }}>

                <ScrollView contentContainerStyle={styles.MainContainer}>

                    <FlatList
                        data={ this.state.dataLayouts }
                        renderItem={({ item, index }) =>
                            <Panel
                                title={item.title}
                                onRowPress={this.onRowPress} >
                                <FlatList
                                    data={ this.state.dataForms }
                                    renderItem={({ item }) =>
                                        (item.layoutID === index + 1) ?

                                            <ListItem
                                                style={styles.FlatListItemStyle}
                                                //onPress={() => this.props.navigation.navigate('Reports')}
                                                title={item.title}
                                                subtitle={item.dateCreated}
                                            />
                                            :null

                                    }
                                    keyExtractor={item => item.orderNo}
                                />
                            </Panel>

                        }
                        keyExtractor={item => item.id}
                        refreshing={this.state.refreshing}

                    />


                </ScrollView>
            </View>
        );
    }
}



const styles = StyleSheet.create({


    activityIndicator: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
        height: 80
    },

    MainContainer: {
        justifyContent: 'center',
        flex: 1,
        // margin: 10,
        paddingTop: (Platform.OS === 'ios') ? 20 : 0,
    },
    FlatListItemStyle: {
        padding: 10,
        //fontSize: 18,
        height: 60
    },
    container: {
        flex: 1,
        backgroundColor: 'white'
    }
});

export default TemplateScreen;