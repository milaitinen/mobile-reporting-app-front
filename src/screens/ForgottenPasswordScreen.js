import React from 'react';
import { StyleSheet, View, Text, TouchableHighlight } from 'react-native';

export default class ForgottenPasswordScreen extends React.Component {

  handleButtonPress = () => {
      this.props.navigation.goBack();
  };

  render() {
      return (
          <View style={styles.container}>
              <TouchableHighlight style={styles.button} onPress={() => this.handleButtonPress()}>
                  <Text style={styles.text}>Forgotten Password Screen</Text>
              </TouchableHighlight>
          </View>
      );
  }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        justifyContent: 'center',
    },
    button: {
        backgroundColor: '#4F6D7A',
        width: 250,
        height: 50,
        justifyContent: 'center',
    },
    text: {
        textAlign: 'center',
        color: '#fff',
    }
});