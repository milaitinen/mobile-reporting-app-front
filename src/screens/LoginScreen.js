import React from 'react';
import { Text, StatusBar } from 'react-native';

import loginStyles from './style/loginStyles';
import { SignInButton } from '../components/Button';
import { Input } from '../components/TextInput';
import { AppBackground } from '../components/AppBackground';

export default class LoginScreen extends React.Component {

    constructor(props)
    {
        super(props);
        this.state = {
            isLoading: true,
            TextInputUser: '',
            TextInputPassword: '',
            TextInputServer: ''
        };
    }

    render() {
        return (
            <AppBackground>
                <StatusBar backgroundColor='#3d4f7c' barStyle='light-content'/>

                <Text style={loginStyles.title}>
                    MR-Application
                </Text>

                <Text style={loginStyles.slogan}>
                    Keep calm and keep reporting
                </Text>

                <Input
                    name={'user'}
                    placeholder='Email'
                    onChangeText={TextInputUser => this.setState({ TextInputUser })}
                />
                <Input
                    name={'lock'}
                    secureTextEntry={true}
                    placeholder='Password'
                    onChangeText={TextInputPassword => this.setState({ TextInputPassword })}
                />
                <Input
                    name={'globe'}
                    placeholder='Server url'
                    onChangeText={TextInputServer => this.setState({ TextInputServer })}
                />

                <SignInButton onPress={() => this.props.navigation.navigate('drawerStack')}>
                    Sign In
                </SignInButton>

                <Text style={loginStyles.copyright}>
                    Copyright Â© Arter Oy 2017
                </Text>
            </AppBackground>
        );
    }
}