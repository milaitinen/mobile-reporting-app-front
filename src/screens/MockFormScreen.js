import React from 'react';
import { View, FlatList, Text, StyleSheet } from 'react-native';

const dummyForms = [...new Array(12)].map((e, i) => `MockForm${i + 1}`);


export default class MockForm extends React.Component {
  static displayName = 'MockForm';

  render() {
      return (
          <View>
              <FlatList
                  data={dummyForms}
                  renderItem={({ item }) => <Text style={styles.FlatListItemStyle}>{item}</Text>}
                  keyExtractor={(item, index) => index}
              />
          </View>
      );
  }
}

const styles = StyleSheet.create({
    FlatListItemStyle: {
        padding: 10,
        fontSize: 15,
        height: 44
    }
});