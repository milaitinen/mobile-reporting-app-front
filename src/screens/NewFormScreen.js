import React from 'react';
import { StyleSheet, View, TextInput, Platform, Alert, Button } from 'react-native';
import { url } from './urlsetting';
import { NavigationActions } from 'react-navigation';
export default class NewFormScreen extends React.Component {


    constructor(props)
    {
        super(props);
        this.state = {
            TextInputName  : '',                                      // Text input is initialized as an empty string.
            layoutID       : props.navigation.state.params.layoutID,  /* LayoutID that is inherited from navigation
                                                                         parameters as stated in TemplateScreen class. */

        };

    }

    // Gets the current date and return it as a string.

    getDate = () => {

        const today = new Date();
        let dd = today.getDate();
        let mm = today.getMonth() + 1; // January is 0!
        const yyyy = today.getFullYear();

        if (dd < 10) {
            dd = '0' + dd;
        }

        if (mm < 10) {
            mm = '0' + mm;
        }

        return yyyy + '-' + mm + '-' + dd;
    }


    // Inserts data to server with a post method.

    InsertDataToServer = () => {

        const date = this.getDate();

        fetch(url + '/users/1/forms', {
            method: 'POST',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            },

            body: JSON.stringify({
                layoutID: this.state.layoutID,
                title: this.state.TextInputName,
                dateCreated: date,
                answers: [
                    {
                        fieldID: 1,
                        answer: "Answer 1"
                    },
                    {
                        fieldID: 2,
                        answer: "Answer 2"
                    }
                ]

            })

        }).then(response => {
            if (response.status === 200) {
                this.props.navigation.state.params.refresh();
                this.props.navigation.dispatch(NavigationActions.back());
                return 'Report added';
            } else {
                return response.status;
            }
        }).then((message) => {
            // Showing response message coming from server after inserting records.
            Alert.alert(message);

        }).catch((error) => {
            console.error(error);
        });

    }
    
    render() {
        return (
            <View style={styles.MainContainer}>
                <TextInput
                    placeholder='Enter Report Name'

                    onChangeText={TextInputName => this.setState({ TextInputName })}

                    underlineColorAndroid='transparent'

                    style={styles.TextInputStyleClass}
                />
                <Button
                    title='Create new report'
                    onPress={ this.InsertDataToServer }
                >
                </Button>
            </View>
        );
    }
}

const styles = StyleSheet.create({
    MainContainer: {
        justifyContent: 'center',
        // alignItems: 'center',
        flex: 1,
        paddingTop: (Platform.OS === 'ios') ? 20 : 0,
    },

    TextInputStyleClass: {
        textAlign: 'center',
        marginBottom: 7,
        height: 40,
        borderWidth: 1,
        borderColor: '#FF5722',
    },
});