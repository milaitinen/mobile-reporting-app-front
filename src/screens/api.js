import { Platform, AsyncStorage, NetInfo } from 'react-native';

import { url } from './urlsetting';

export const login = (username, password) => {
    return fetch(`${url}/login`, {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            'username': username,
            'password': password,
        })
    }).then(response => {
        return JSON.parse(response._bodyInit).token;
    }).catch(err => alert(err));
};

// Send a new report to the server, along with the username and token.
export const createNewReport = (username, report, token) => {
    return fetch(`${url}/users/${username}/reports`, {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${token}`,
        },
        body: JSON.stringify(report)
    });
};

export const fetchFieldsByTemplateID = (username, templateID, token) => {
    return isNetworkConnected()
        .then((isConnected) => {
            if (!isConnected) { return fetchLocalFieldsByTemplateID(username, templateID); }
            return fetchRemoteFieldsByTemplateID(username, templateID, token);
        })
        .then((fieldsByTemplateID) => {
            saveData(`${url}/users/${username}/templates/${templateID}/fields`, fieldsByTemplateID);
            return fieldsByTemplateID;
        });
};

const fetchLocalFieldsByTemplateID = (username, templateID) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates/${templateID}/fields`)
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

export const fetchRemoteFieldsByTemplateID = (username, templateID, token) => {
    return (
        fetch(`${url}/users/${username}/templates/${templateID}/fields`, {
            headers: {
                'Authorization': `Bearer ${token}`,
            }
        })
            .then(response => {
                return response.json();
            })
    );
};

// used to store drafts
export const saveDraft = (username, templateID, draft) => {
    fetchDraftsByTemplateID(username, templateID)
        .then((drafts) => {
            // see if there is already a draft with the same id
            const draftIndex = drafts.findIndex(x => x.id === draft.id);

            if (draftIndex < 0) {
                drafts.push(draft);
            } else {
                drafts[draftIndex] = draft;
            }
            // give each draft a unique, negative id
            drafts.map((draft, i) => draft.id = -Math.abs(i + 1));
            saveData(`${url}/users/${username}/templates/${templateID}`, drafts);

            return (drafts[drafts.length - 1].id);
        });
};

export const removeDraft =  (username, templateID, draftID) => {
    fetchDraftsByTemplateID(username, templateID)
        .then((drafts) => {
            const draftRemoved = drafts.filter(draft => draft.id !== draftID);
            saveData(`${url}/users/${username}/templates/${templateID}`, draftRemoved);
        });
};

export const fetchDraftsByTemplateID = (username, templateID) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates/${templateID}`) // NOTE: this url is just a key
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

// fetch reports that are only locally saved
export const fetchLocalReports = (username, templateID) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates/${templateID}`) // NOTE: this url is just a key
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

// for future use?
export const fetchFieldsByReportID = (username, templateID, reportID, token) => {
    return isNetworkConnected()
        .then((isConnected) => {
            if (!isConnected || reportID === 0) { return fetchLocalFieldsByReportID(username, templateID, reportID); }
            return fetchRemoteFieldsByReportID(username, templateID, reportID, token);
        })
        .then((fieldsByReportID) => {
            saveData(`${url}/users/${username}/templates/${templateID}/reports/${reportID}/fields`, fieldsByReportID);
            return fieldsByReportID;
        });
};

// not yet in use
export const fetchLocalFieldsByReportID = (username, templateID, reportID) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates/${templateID}/reports/${reportID}/fields`)
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

// not yet in use
const fetchRemoteFieldsByReportID = (username, templateID, reportID, token) => {
    return (
        fetch(`${url}/users/${username}/templates/${templateID}/reports/${reportID}/fields`, {
            headers: {
                'Authorization': `Bearer ${token}`,
            }
        })
            .then(response => {
                return response.json();
            })
    );
};


/*
 Fetch templates from the server or ASyncStorage, depending on the availability of internet connection.
 Fetch templates that the user has rights to. */
export const fetchTemplatesByUsername = (username, token) => {
    return isNetworkConnected()
        .then((isConnected) => {
            if (!isConnected) { return fetchLocalTemplatesByUsername(username); }
            return fetchRemoteTemplatesByUsername(username, token);
        })
        .then((templates) => {
            saveData(`${url}/users/${username}/templates`, templates);
            return templates;
        });
};

/*
 Fetch all templates from ASyncStorage in case there is no internet connection.
 If no data has been stored an empty value will be returned. */
const fetchLocalTemplatesByUsername = (username) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates`)
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

// Fetch all templates from the server
const fetchRemoteTemplatesByUsername = (username, token) => {
    return (
        fetch(`${url}/users/${username}/templates`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        })
            .then(response => {
                if (response._bodyText === '') response._bodyText = '[]';
                return response.json();
            })
    );
};

/* Fetch reports by their templateID from the server if online, ASyncStorage otherwise */
export const fetchReportsByTemplateID = (username, templateID, token) => {
    return isNetworkConnected()
        .then((isConnected) => {
            if (!isConnected) { return fetchLocalReportsByTemplateID(username, templateID); }
            return fetchRemoteReportsByTemplateID(username, templateID, token);
        })
        .then((reports) => {
            saveData(`${url}/users/${username}/templates/${templateID}/reports?sort=-datecreated`, reports);
            return reports;
        });
};

/* Fetch Reports by TemplateID from ASyncStorage in case there is no internet connection.
   If no data has been stored an empty value will be returned. */
const fetchLocalReportsByTemplateID = (username, templateID) => {
    return AsyncStorage.getItem(`${url}/users/${username}/templates/${templateID}/reports?sort=-datecreated`)
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

// Fetch reports by templateID from the server
const fetchRemoteReportsByTemplateID = (username, templateID, token) => {
    return (
        fetch(`${url}/users/${username}/templates/${templateID}/reports?sort=-datecreated`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        })
            .then(response => {
                return response.json();
            })
    );
};

/* Returns all the reports made by the user specified by the username. */
export const fetchReportsByUsername = (username, token) => {
    return isNetworkConnected()
        .then((isConnected) => {
            if (!isConnected) { return fetchLocalReportsByUsername(username); }
            return fetchRemoteReportsByUsername(username, token);
        })
        .then((reports) => {
            saveData(`${url}/users/${username}/reports`, reports);
            return reports;
        });
};

const fetchLocalReportsByUsername = (username) => {
    return AsyncStorage.getItem(`${url}/users/${username}/reports`)
        .then(data => {
            if (data !== null) {
                return JSON.parse(data);
            } else {
                return [];
            }
        });
};

const fetchRemoteReportsByUsername = (username, token) => {
    return (
        fetch(`${url}/users/${username}/reports`, {
            headers: {
                'Authorization': `Bearer ${token}`
            }
        })
            .then(response => {
                return response.json();
            })
    );
};

/* Store data (layouts, reports - depending on the url) to ASyncStorage, a simple key-value storage system global to the app.
   Keys and values are stored as a string. */
const saveData = (dataUrl, data) => {
    AsyncStorage.setItem(dataUrl, JSON.stringify(data));
};

/*
const removeData = (dataUrl) => {
    try {
        AsyncStorage.removeItem(dataUrl);
    } catch (error) {
        console.error(error);
    }
};
*/

// Necessary because of a bug on iOS https://github.com/facebook/react-native/issues/8615#issuecomment-287977178
export function isNetworkConnected() {
    if (Platform.OS === 'ios') {
        return new Promise(resolve => {
            const handleFirstConnectivityChangeIOS = isConnected => {
                NetInfo.isConnected.removeEventListener('connectionChange', handleFirstConnectivityChangeIOS);
                resolve(isConnected);
            };
            NetInfo.isConnected.addEventListener('connectionChange', handleFirstConnectivityChangeIOS);
        });
    }

    return NetInfo.isConnected.fetch();
}
