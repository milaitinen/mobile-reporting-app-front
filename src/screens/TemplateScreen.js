import React, { Component } from 'react';
import {
    View,
    FlatList,
    ActivityIndicator,
    ScrollView,
    StatusBar,
} from 'react-native';
import { connect } from 'react-redux';
import moment from 'moment';

import templateScreenStyles from './style/templateScreenStyles';
import { Layout } from '../components/Layout';
import { AppBackground } from '../components/AppBackground';
import { ReportSearchBar } from '../components/ReportSearchBar';
import { fetchReportsByTemplateID, fetchTemplatesByUsername, fetchDraftsByTemplateID } from './api';
import { storeTemplates } from '../redux/actions/templates';
import { storeReportsByTemplateID, storeDraftByTemplateID } from '../redux/actions/reports';
import { createReport } from '../redux/actions/newReport';
import { preview } from '../redux/actions/preview';
import userReducer from '../redux/reducers/user';


// "export" necessary in order to test component without Redux store
export class TemplateScreen extends Component {

    constructor(props)
    {
        super(props);
        this.state = {
            isLoading       : true,     // Checks whether the app is loading or not.
            refreshing      : false,    // Checks whether the app and its data is refreshing or not.
            scrollEnabled   : true,     // Checks whether the template screen is scrollable or not.
            renderFooter    : false     // If true, empty space is rendered after the last template. This is set to true while a template is opened.
        };
    }

    /*
    handleBackPress = () => {
        if (this.backPressed && this.backPressed > 0) {
            this.props.navigator.popToRoot({ animated: false });
            return false;
        }

        this.backPressed = 1;
        this.props.navigator.showSnackbar({
            text: 'Press one more time to exit',
            duration: 'long',
        });
        return true;
    }
    */

    /*
     componentDidMount() is invoked immediately after the component is mounted. Initialization that requires
     DOM nodes happens here. The function calls getTemplates which loads data from a remote url,
     and instantiates the network request.
    */
    componentDidMount() {
        // TEMPORARY: not sure if this is the best solution. Current version fixes a bug (related to logging in)
        if (this.props.username !== userReducer.username) {
            this.getTemplatesAndReports();
        } else {
            this.setState({ refreshing: false, isLoading: false });
        }

    }

    /*
     Fetches the data from the server in two parts.
     1) Fetches the templates from the server
     2) Fetches the reports under their specific template by making a separate fetch request using
        Promise.all. After all the promises have been fetched, the function updates the state
        of reportsByTemplates, and sets isLoading and refreshing to false.
    */
    getTemplatesAndReports = () => {
        const { username, token } = this.props;

        fetchTemplatesByUsername(username, token)
            .then(responseJson => {
                if (responseJson.length < 1) {  // handle situations where there are no templates
                    this.setState({ refreshing: false, isLoading: false });
                } else {
                    this.props.dispatch(storeTemplates(responseJson));
                }
            })
            .then(() => {
                const reportsByTemplateID = Object.keys(this.props.templates)
                    .map((templateID) => fetchReportsByTemplateID(username, templateID, token));

                Promise.all(reportsByTemplateID)
                    .then(data => this.props.dispatch(storeReportsByTemplateID(data)))
                    .then(() => this.getDrafts())
                    .catch(err => console.error(err));
            })
            .catch(error => console.error(error))
            .done();
    };

    getDrafts = () => {
        const { templates, username } = this.props;

        Object.keys(templates).forEach((templateID) => {
            fetchDraftsByTemplateID(username, templateID)
                .then((drafts) => {
                    if (drafts.length !== 0) {
                        drafts.forEach(draft => this.props.dispatch(storeDraftByTemplateID(templateID, draft)));
                    }
                })
                .then(() => this.setState({ refreshing: false, isLoading: false }))
                .catch(err => console.error(err));
        });
    };

    // Handler function for refreshing the data and refetching.
    handleRefresh = () => {
        this.setState({ refreshing: true, }, () => { this.getTemplatesAndReports(); });
        this.setState({ isLoading: true });
    };

    // Determines whether this screen is scrollable or not.
    setScrollEnabled = (bool) => {
        this.setState({ scrollEnabled : bool });
    };

    /*
     Determines whether empty space is rendered after the last template.
     Without this function it wouldn't be possible to autoscroll to the last templates.
     */
    setRenderFooter = (bool) => {
        this.setState({ renderFooter: bool });
    };

    /*
     Function that passes navigation props and navigates to NewReportScreen.
     This makes it possible for the Layout component to navigate.
     Also passes the refresh function and the specific TemplateID so that the
     app knows to which template the new report has to be added.
    */
    createNew = (templateID, isEditable) => {
        if (isEditable) {
            this.props.dispatch(createReport(templateID, moment().format('YYYY-MM-DD')));
            // this.setState({ isLoading: true });
            this.props.navigation.navigate('NewReport', { refresh: this.handleRefresh, isEditable: isEditable });
        }
        else {
            this.props.dispatch(preview(templateID));
            // this.setState({ isLoading: true });
            this.props.navigation.navigate('Preview', { refresh: this.handleRefresh,  isEditable: isEditable });
        }
        //this.setState({ isLoading: true }); TODO fix backhandler issue in NewReport, Preview, and ReportScreen and uncomment this
    };

    viewReport = (templateID, reportID, title) => {
        //this.setState({ isLoading: true }); TODO same problem as above
        this.props.navigation.navigate('Report',
            { refresh: this.handleRefresh, templateID: templateID, reportID: reportID, title: title });
    };

    render() {
        if (this.state.isLoading) {
            return (
                <AppBackground>
                    <ActivityIndicator
                        animating={this.state.animating}
                        style={[templateScreenStyles.activityIndicator, { height: 80 }]}
                        size='large'
                        color={'#88daf2'}
                    />
                </AppBackground>
            );
        }

        const { reports, templates } = this.props;

        return (
            <AppBackground>
                <View style={templateScreenStyles.viewContainer}>
                    <StatusBar
                        backgroundColor={templateScreenStyles.statusBar}
                        barStyle='light-content'
                    />
                    {/* At the moment this doesn't do anything.*/}
                    <ReportSearchBar/>
                    <ScrollView contentContainerStyle={templateScreenStyles.scrollView}>
                        <FlatList
                            ref={(c) => this._flatList = c}
                            scrollEnabled={this.state.scrollEnabled}
                            data={ Object.values(templates) }
                            renderItem={({ item, index }) =>
                                <Layout
                                    title={item.title}
                                    moveToTop={(viewPosition = 0) => this._flatList.scrollToIndex({ animated: true, index: index, viewPosition: viewPosition })}
                                    setTemplateScreenScrollEnabled={this.setScrollEnabled}
                                    setTemplateScreenRenderFooter={this.setRenderFooter}
                                    createNew={this.createNew}
                                    viewReport={this.viewReport}
                                    nofReports={(reports[item.id]) ? (reports[item.id]).length : 0}
                                    templateID={item.id}
                                    data={reports[item.id]}
                                />
                            }
                            ListFooterComponent={
                                (this.state.renderFooter) && <View style={{ height: 500 }}/>
                            }
                            keyExtractor={item => item.id}
                            refreshing={this.state.refreshing}
                        />
                    </ScrollView>
                </View>
            </AppBackground>
        );
    }
}

// maps redux state to component props. Object that is returned can be accessed via 'this.props' e.g. this.props.email
const mapStateToProps = (state) => {
    const username = state.user.username;
    const templates = state.templates;
    const reports = state.reports;
    const token = state.user.token;
    return {
        username,
        templates,
        reports,
        token
    };
};

export default connect(mapStateToProps)(TemplateScreen);
