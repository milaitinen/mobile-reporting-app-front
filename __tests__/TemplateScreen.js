import 'react-native';
import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { ActivityIndicator } from 'react-native';
import renderer from 'react-test-renderer';

import { TemplateScreen } from '../src/screens/TemplateScreen';

configure({ adapter: new Adapter() });

it('renders correctly', () => {
    const tree = renderer.create(
        <TemplateScreen />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});


describe('<TemplateScreen />', () => {
    describe('isLoading', () => {
        it('should render a <ActivityIndicator /> if true', () => {
            const templateScreen = shallow(<TemplateScreen />);
            templateScreen.setState({ isLoading: true });
            expect(templateScreen.find(ActivityIndicator).length).toBe(1);
        });
    });
});
