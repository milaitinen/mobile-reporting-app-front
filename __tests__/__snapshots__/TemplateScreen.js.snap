// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateScreen renders correctly 1`] = `
<View
  colors={
    Array [
      undefined,
      undefined,
      undefined,
    ]
  }
  style={undefined}
>
  <View
    style={undefined}
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#393e42",
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "borderTopColor": "#000",
            "borderTopWidth": 1,
          },
          Object {
            "backgroundColor": "#e1e8ee",
            "borderBottomColor": "#e1e1e1",
            "borderTopColor": "#e1e1e1",
          },
          undefined,
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        placeholder="Search.."
        placeholderTextColor="#86939e"
        style={
          Array [
            Object {
              "backgroundColor": "#303337",
              "borderRadius": 3,
              "color": "#86939e",
              "fontSize": 17.5,
              "height": 30,
              "margin": 8,
              "overflow": "hidden",
              "paddingLeft": 26,
              "paddingRight": 19,
            },
            Object {
              "backgroundColor": "#bdc6cf",
            },
            false,
            false,
            undefined,
            undefined,
            false,
          ]
        }
        underlineColorAndroid="transparent"
      />
      <Text
        accessible={true}
        allowFontScaling={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#86939e",
              "fontSize": 16,
            },
            Array [
              Object {
                "backgroundColor": "transparent",
                "position": "absolute",
                "top": 15.5,
              },
              Object {
                "left": 16,
              },
              undefined,
            ],
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        î¢¶
      </Text>
    </View>
    <RCTScrollView
      contentContainerStyle={undefined}
    >
      <View>
        <RCTScrollView
          ListFooterComponent={false}
          data={Array []}
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={false}
          initialNumToRender={10}
          keyExtractor={[Function]}
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          refreshing={false}
          renderItem={[Function]}
          scrollEnabled={true}
          scrollEventThrottle={50}
          stickyHeaderIndices={Array []}
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View />
        </RCTScrollView>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
